File: apps/rabbitci_core/web/models/project.ex
defmodule RabbitCICore.Project do
  use RabbitCICore.Web, :model

  alias RabbitCICore.Branch
  alias RabbitCICore.SSHKey

  schema "projects" do
    field :name, :string
    field :repo, :string
    field :webhook_secret, :string

    has_many :branches, Branch
    has_one :ssh_key, SSHKey

    timestamps
  end

  @doc """
  Creates a changeset based on the `model` and `params`.

  If no params are provided, an invalid changeset is returned
  with no validation performed.
  """
  def changeset(model, params \\ %{}) do
    model
    |> cast(params, ~w(name repo), ~w(webhook_secret))
    |> validate_format(:name, ~r/^[^\/]+\/[^\/]+$/)
    |> unique_constraint(:name)
    |> unique_constraint(:repo)
  end
end

File: frontend/app/models/project.js
import DS from 'ember-data';

export default DS.Model.extend({
  name: DS.attr('string'),
  repo: DS.attr('string'),
  branches: DS.hasMany('branch'),

  owner: Ember.computed('name', function() {
    if (this.get('name')) {
      return this.get('name').split('/')[0];
    } else return undefined;
  }),
  repo_name: Ember.computed('name', function() {
    if (this.get('name')) {
      return this.get('name').split('/')[1];
    } else return undefined;
  })
});
